index/postings.go

```diff
@@ -117,9 +122,61 @@ func (p *memPostings) ensureOrder() {
 	p.ordered = true
 }
 
-// add adds a document to the index. The caller has to ensure that no
-// term argument appears twice.
-func (p *memPostings) add(id uint64, lset labels.Labels) {
+// Delete removes all ids in the given map from the postings lists.
+func (p *MemPostings) Delete(deleted map[uint64]struct{}) {
+	var keys []labels.Label
+
+	p.mtx.RLock()
+	for l := range p.m {
+		keys = append(keys, l)
+	}
+	p.mtx.RUnlock()
+
+	for _, l := range keys {
+		p.mtx.Lock()
+
+		found := false
+		for _, id := range p.m[l] {
+			if _, ok := deleted[id]; ok {
+				found = true
```

Comment:
You are not doing `repl = append(repl, id)` here to lock as short time as possible, right?

Reply:
Yea and to not reallocate every single postings list on GC even if it wasn't changed.